<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Transitional//EN" "http://www.w3.org/TR/xhtml1/DTD/xhtml1-transitional.dtd">
<html xmlns="http://www.w3.org/1999/xhtml">
<head>
  <meta http-equiv="Content-Type" content="text/html; charset=utf-8" />
  <meta http-equiv="Content-Style-Type" content="text/css" />
  <meta name="generator" content="pandoc" />
  <title></title>
  <style type="text/css">code{white-space: pre;}</style>
</head>
<body>
<h1 id="varan-gie-an-igv-extension-for-the-annotation-of-genomic-intervals.">VARAN-GIE: An IGV extension for the annotation of genomic intervals.</h1>
<h2 id="introduction">Introduction</h2>
<p>VARAN-GIE (aka &quot;VARAN&quot;) is a <em>branch</em> of <a href="http://software.broadinstitute.org/software/igv/">Broad's popular Integrative Genomics Viewer (IGV)</a>. It <em>extends</em> the IGV by adding <strong>functionality to create, edit and annotate sets of genomic intervals</strong>. By this, VARAN supports an integrative approach to viewing and annotating large genomic data sets.</p>
<p>Core functionality of VARAN includes:</p>
<ul>
<li><strong>Simple and intuitive editing and annotation of genomic interval sets</strong>
<ul>
<li>Editing functionality includes: add/delete/clip/merge.</li>
<li>VARAN supports authoring of complex annotation sets containing potentially overlapping intervals on multiple layers.</li>
</ul></li>
<li><strong>Simple merging/curation of multiple genomic interval sets (e.g., a set of BED files)</strong>
<ul>
<li>Intervals can be organized in multiple custom layers to enable the authoring of complex annotation sets</li>
</ul></li>
<li><strong>Interval annotation capabilities</strong>
<ul>
<li>VARAN supports editing of extended BED features such as score, strand and color and arbitrary custom attributes</li>
</ul></li>
<li><strong>Comprehensive navigational capabilities</strong>
<ul>
<li>E.g., zoom to intervals/endpoints, forward/back in interval list, etc.</li>
</ul></li>
<li><strong>Session management: users can easily switch between different VARAN annotation data sets/versions.</strong>
<ul>
<li>Datasets can be organized in custom categories by the user</li>
</ul></li>
<li><strong>Export functionalites</strong>
<ul>
<li>Interval sets can be exporteed in spreadsheet-compatible formats (CSV/TSV) or as UCSC bed tracks</li>
<li>Whole datasets can be exported as zipped &quot;bundles&quot; and exchanged with other researchers (referenced file paths can be mapped to new locations at import)</li>
</ul></li>
<li><strong>Autosave, versioning and &quot;undo&quot; functionalities</strong></li>
</ul>
<p>VARAN was developed at the <a href="http://science.ccri.at/">Children's Cancer Research Institute (CCRI)</a> where we have used it in several projects (e.g., for curating CNV calls coming out of bioinformatics pipelines or for visualizing and editing data in a gene panel project).</p>
<h2 id="screenshot">Screenshot</h2>
<div class="figure">
<img src="README.GIE.png" title="VARAN screenshot" alt="VARAN screenshot showing 3 windows:" /><p class="caption">VARAN screenshot showing 3 windows:</p>
</div>
<ul>
<li>The main IGV window in the background shows the data for two SNP array datasets (2A and 2B) that were created by a bioinformatics pipeline that extracts and analyses BAF and LogR signals and predicts CNV segments from these data. The fragmented output from this pipeline (&quot;Segments&quot; tracks) was curated using VARAN's interval editing functionality, resulting in two segments (a gain and a loss) for datasets 2A. The segements are indicated by the blue and red bars above the data tracks and in the &quot;2A&quot; track.</li>
<li>The upper window in the foreground shows all datasets in the &quot;Example&quot; category known to the running VARAN instance. The first dataset (&quot;p003&quot;) has 2 different versions, the second one (&quot;p001&quot;) is currently loaded.</li>
<li>The lower window in the foreground shows the two intervals that are currently stored in layer &quot;2A&quot;</li>
</ul>
<h2 id="possible-application-scenarios">Possible application scenarios</h2>
<ul>
<li>Manual curation of genomic annotations, e.g., gene lists, CNV or SV calls, novel transcripts</li>
<li>Merging and curation of multiple BED files</li>
<li>De-novo authoring of <em>multi-layer</em> interval sets</li>
<li>etc.</li>
</ul>
<h2 id="varan-key-mappings">VARAN Key Mappings</h2>
<table>
<thead>
<tr class="header">
<th align="left">Key</th>
<th align="left">Mapping</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td align="left">CTRL-ALT-A</td>
<td align="left">Add region (hold CTRL to merge new region with all overlapping)</td>
</tr>
<tr class="even">
<td align="left">CTRL-ALT-C</td>
<td align="left">Clip region</td>
</tr>
<tr class="odd">
<td align="left">CTRL-ALT-M</td>
<td align="left">Merge visible/selected regions</td>
</tr>
<tr class="even">
<td align="left">CTRL-ALT-D</td>
<td align="left">Delete visible/selected regions</td>
</tr>
<tr class="odd">
<td align="left">CTRL-ALT-R</td>
<td align="left">Save and reload BED track</td>
</tr>
<tr class="even">
<td align="left">CTRL-F</td>
<td align="left">Next (downstream) region</td>
</tr>
<tr class="odd">
<td align="left">CTRL-B</td>
<td align="left">Previous (upstream) region</td>
</tr>
<tr class="even">
<td align="left">CTRL-Z</td>
<td align="left">Undo last action</td>
</tr>
<tr class="odd">
<td align="left">CTRL-C</td>
<td align="left">Copy region coordinates</td>
</tr>
<tr class="even">
<td align="left">CTRL-H</td>
<td align="left">Show/hide reference lines</td>
</tr>
<tr class="odd">
<td align="left">CTRL-#</td>
<td align="left">Switch to layer (#=1..9)</td>
</tr>
</tbody>
</table>
<h2 id="igv-modifications-and-bugfixes">IGV modifications and bugfixes</h2>
<p>VARAN also includes several bugfixes and other extensions of the branched IGV version: * &quot;Load session&quot; allows to extend/merge current session. * Overlay track handling: track context menu allows to set windowing function and height for overlay tracks; configured renderer of overlay-tracks is kept when session is saved * Proper UTF-8 encoding of session XML. * Many smaller bugfixes (see commit logs for details)</p>
<h2 id="roadmap-for-next-version">Roadmap for next version</h2>
<ul>
<li>Enable filtering interval lists by name, chrom, etc.</li>
<li>Improve undo to also support interval edits (currently only additions/removals of intervals are tracked)</li>
<li>Improve import path mapping; include http(s) URIs</li>
</ul>
</body>
</html>
